### PR TITLE
[mqtt] Discovery services shall not unsubscribe unless they have already subscribed

### PR DESCRIPTION
**Pull request by andrewfg**
The MQTT AbstractBrokerHandler produces erroneous logger warnings when OH is being shutdown for the EspMilight, HomeAssistant, and Homie, discovery services. See #10562 

I think that the problem is because of some or all of the following issues: 
1. The three respective discovery participant classes extend AbstractMQTTDiscovery, and that class may be calling `getDiscoveryService().unsubscribe()` although it might not already have called `getDiscoveryService().subscribe()`.
2. The MqttBrokerHandlerFactory discoveryTopics Map allows duplicate entries of the same listener in each topic's listenerList.
3. The MqttBrokerHandlerFactory subscribe method would re-subscribe the given listener/topic even if already subscribed.
4. The MqttBrokerHandlerFactory unsubscribe method would unsubscribe ALL the topics (even if the listener was not in the listenerList for that topic) -- **_this IMHO is the most critical error in the current code._**

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

### CODE REVIEW

**Comment by andrewfg** [comment]
I found a way to install all the dependencies so I was able to test it :)
=> Removing the "additional testing required" flag

**Comment by andrewfg** [comment]
> shouldn't we rename the variable to 'listenerSet'

@jochen314 done.

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java [hunk 1]
```diff
@@ -48,11 +48,34 @@
 
     private @Nullable ScheduledFuture<?> scheduledStop;
 
+    private boolean isSubscribed;
```
Comment (jochen314) [inline bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java:51]: To make thsi really thread safe, this should be an `AtomicBoolean`.
Then can use `getAndSet()` to set and test the flag in a thread safe way.

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java [hunk 2]
```diff
@@ -48,11 +48,34 @@
...
         this.subscribeTopic = baseTopic;
         this.timeout = timeout;
+        isSubscribed = false;
+    }
+
+    /**
+     * Only subscribe if we were not already subscribed
+     */
+    private void subscribe() {
+        if (!isSubscribed) {
```
Comment (jochen314) [inline bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java:65]:
```
if (!isSubscribed.getAndSet(true)) {
    getDiscoveryService().subscribe(this, subscribeTopic);
}
````

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java [hunk 3]
```diff
@@ -48,11 +48,34 @@
...
+            getDiscoveryService().subscribe(this, subscribeTopic);
+            isSubscribed = true;
+        }
+    }
+
+    /**
+     * Only unsubscribe if we were already subscribed
+     */
+    private void unSubscribe() {
+        if (isSubscribed) {
```
Comment (jochen314) [inline bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java:75]: if (isSubscribed.getAndSet(false)) {
    getDiscoveryService().unsubscribe(this);
}

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java [hunk 4]
```diff
@@ -54,11 +53,22 @@
...
             .collect(Collectors.toSet());
+
     private final Logger logger = LoggerFactory.getLogger(MqttBrokerHandlerFactory.class);
-    protected final Map<String, List<MQTTTopicDiscoveryParticipant>> discoveryTopics = new HashMap<>();
+
+    /**
+     * This Map provides a lookup between a Topic string (key) and a Set of MQTTTopicDiscoveryParticipants (value),
+     * where the Set itself is a list of participants which are subscribed to the respective Topic.
+     */
+    protected final Map<String, Set<MQTTTopicDiscoveryParticipant>> discoveryTopics = new HashMap<>();
```
Comment (jochen314) [inline bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java:67]: To improve thead safety, this should become a `ConcurrentHashMap`

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java [hunk 5]
```diff
@@ -111,24 +122,32 @@ protected void createdHandler(AbstractBrokerHandler handler) {
...
+     * list, and registers itself and the respective topic with all the known brokers.
      */
     @Override
+    @SuppressWarnings("null")
     public void subscribe(MQTTTopicDiscoveryParticipant listener, String topic) {
-        List<MQTTTopicDiscoveryParticipant> listenerList = discoveryTopics.computeIfAbsent(topic,
-                t -> new ArrayList<>());
-        listenerList.add(listener);
-        handlers.forEach(broker -> broker.registerDiscoveryListener(listener, topic));
+        Set<MQTTTopicDiscoveryParticipant> listeners = discoveryTopics.computeIfAbsent(topic, t -> new HashSet<>());
```
Comment (jochen314) [inline bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java:132]: for thread safety, we should make this `ConcurrentHashMap.newKeySet()`

**Review by jochen314 — COMMENTED** [review]
If we now improve the thread saftey, it will look good to me

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java (hunk 1 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java:51]: Ok.

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java (hunk 2 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java:65]: Ok.

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java (hunk 3 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java:75]: Ok.

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java (hunk 4 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java:67]: Ok.

File: bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java (hunk 5 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/internal/MqttBrokerHandlerFactory.java:132]: I changed the code to the following. I am not familiar with the constructor, and so wonder if the `<>` needs to be specifically defined ??

```
        Set<MQTTTopicDiscoveryParticipant> listeners = discoveryTopics.computeIfAbsent(topic,
                t -> ConcurrentHashMap.newKeySet());
```

**Review by jochen314 — APPROVED** [review]
LGTM

**Comment by andrewfg** [comment]
PING!

**Comment by andrewfg** [comment]
PING!!

**Comment by Skinah** [comment]
LGTM

As Code owner for mqtt.espmilighthub this has been a minor annoyance in the logs but have not seen any issues. From the description of this, it may be increasing system load if a listener can get subscribed multiple times, so thanks for making the changes.

**Review by kaikreuzer — APPROVED** [review]
Thanks @andrewfg!

**Pull request merged**